File: templates/nodejs/v/latest/src/strato_task.js
import { execSync } from 'child_process';
import fs from 'fs';
import { handler as taskHandler } from './task.js'; // Ensure correct path and export

// Export handler as an ES module export
export const handler = async (event = {}, context = {}) => {
  const preScriptPath = '/var/task/pre.sh';
  const postScriptPath = '/var/task/post.sh';
  const envFilePath = '/tmp/strato_env.json';

  try {
    if (fs.existsSync(preScriptPath)) {
      console.log('Running pre.sh...');
      fs.writeFileSync(envFilePath, JSON.stringify(event), 'utf-8');

      execSync(`${preScriptPath}`, { stdio: 'inherit' });

      const envData = fs.readFileSync(envFilePath, 'utf-8');
      const envVars = JSON.parse(envData);
      Object.keys(envVars).forEach((key) => {
        console.log(`Setting environment variable: ${key}=${envVars[key]}`);
        process.env[key] = envVars[key];
      });
    } else {
      console.log('pre.sh not found, skipping.');
    }

    const response = await taskHandler(event, context); // Call the imported function

    if (fs.existsSync(postScriptPath)) {
      console.log('Running post.sh...');
      execSync(postScriptPath, { stdio: 'inherit' });
    } else {
      console.log('post.sh not found, skipping.');
    }

    return response;
  } catch (error) {
    console.error('Error occurred:', error);

    if (fs.existsSync(postScriptPath)) {
      console.log('Running post.sh...');
      execSync(postScriptPath, { stdio: 'inherit' });
    } else {
      console.log('post.sh not found, skipping.');
    }

    throw error;
  }
};

File: templates/python/v/latest/src/strato_task.py
# strato_task.py

import os
import subprocess
import json
import sys
from task import task_handler


def run_script(script_path):
    """Run a shell script if it exists."""
    if os.path.exists(script_path):
        print(f'Running {script_path}...')
        subprocess.run(["bash", script_path], check=True)
    else:
        print(f'{script_path} not found, skipping.')


def handler(event=None, context=None):
    """Main task handler that executes pre and post scripts."""
    if event is None:
        event = {}

    try:
        env_file = '/tmp/strato_env.json'
        with open(env_file, 'w') as f: json.dump(event, f)

        run_script('/var/task/pre.sh')

        with open(env_file, 'r') as f:
            env_vars = json.load(f)
            for key, value in env_vars.items():
                os.environ[key] = str(value)

        # Execute main task logic
        response = task_handler()
        print("Task Response:", response)

        # Execute post.sh if it exists
        run_script('/var/task/post.sh')

        return response
    except Exception as error:
        print('Error occurred:', error)
        # Execute post.sh if it exists
        run_script('/var/task/post.sh')
        raise


def parse_args(args):
    """Parse command-line arguments as JSON."""
    if len(args) > 1:
        try:
            event = json.loads(args[1])
        except json.JSONDecodeError as e:
            print('Failed to parse arguments as JSON:', e)
            event = {}
    else:
        event = {}
    return event


if __name__ == '__main__':
    event = parse_args(sys.argv)
    handler(event)
